maup/repair.py

```diff
@@ -1,27 +1,1217 @@
+    that they came from, and assign the pieces inside D to D.  (This is 
+    equivalent to taking the difference of each polygon minus D, except that
+    it works cleanly and reliably!)
+
+(3) For each of the polygons whose boundary (now) contains an arc of D's boundary
+    circle, create a "pie wedge" inside D subtending this arc, and join it to    
+    the polygon that it shares a boundary with.
+    
 """
-Some of these functions are based on the functions in Mary Barker's
-check_shapefile_connectivity.py script in @gerrymandr/Preprocessing.
+
+
+# Some useful short functions:
+
+# Count the number of connected components of a shapely object:
+def num_components(geom):
+    if geom.is_empty:
+        return 0
+    elif (
+        (geom.geom_type == "Polygon")
+        or (geom.geom_type == "Point")
+        or (geom.geom_type == "LineString")
+    ):
+        return 1
+    elif (
+        (geom.geom_type == "MultiPolygon")
+        or (geom.geom_type == "MultiLineString")
+        or (geom.geom_type == "GeometryCollection")
+    ):
+        return len(geom.geoms)
+
+
 """
+MAIN FUNCTION
+"""
+
+
+def autorepair(
+    geometries0_df, fill_gaps=True, min_rook_length=None, holes_to_keep=None
+):
+
+    if isinstance(geometries0_df, geopandas.GeoSeries):
+        geometries_df = geopandas.GeoDataFrame(geometry=geometries0_df)
+
+    elif isinstance(geometries0_df, geopandas.GeoDataFrame):
+        geometries_df = geometries0_df.copy()
+
+    else:
+        raise TypeError(f"Input must be a geopandas GeoSeries or GeoDataFrame!")
+
+    geometries_df = geometries0_df.copy()
+
+    # Ensure that geometries are 2-D and not 3-D!
+    for i in geometries_df.index:
+        geometries_df["geometry"][i] = shapely.wkb.loads(
+            shapely.wkb.dumps(geometries_df["geometry"][i], output_dimension=2)
+        )
+
+    # Warn if crs is geographic:
+    if geometries_df.crs is not None:
+        if geometries_df.crs.is_geographic:
+            warnings.warn(
+                "Geometry is in a geographic CRS. Results from 'length' are likely incorrect. Use 'GeoSeries.to_crs()' to re-project geometries to a projected CRS before this operation."
+            )
+
+    # If holes_to_keep and/or overlaps_to_keep are present, require them to have the
+    # same CRS as the main shapefile.
+    if holes_to_keep is not None:
+        if holes_to_keep.crs != geometries_df.crs:
+            raise Exception(
+                "holes_to_keep must be in the same CRS as the shapefile being repaired."
+            )
+
+    # Construct data about overlaps of all orders, plus holes.
+    overlap_tower, holes_df = building_blocks(geometries_df)
+
+    # Use data from the overlap tower to rebuild precincts with no overlaps.
+    print("Resolving overlaps...")
+    reconstructed_df = reconstruct_from_overlap_tower(geometries_df, overlap_tower)
+
+    # Use data about the holes to fill holes if desired.
+    if fill_gaps == True:
```

Comment:
Minor style tip: it's basically never necessary to compare to `True` like this. This can be replaced with
```suggestion
if fill_gaps:
```